errors.go

```diff
@@ -0,0 +1,164 @@
+// and here: http://jsonapi.org/format/#error-objects
+//
+// To use this function in your code effectively, make your package's error types implement some
+// or all of the error compatibility interfaces (`ErrorTitleCompatible`, `ErrorDetailCompatible`,
+// ...). Doing so gives you direct control over what will appear in the serialied error object's
+// fields. Without implementing these interfaces, the `title` & `detail` fields will be populated
+// with derived data about the error.
+func MarshalErrors(w io.Writer, errs []error) error {
+	// Serialize the given errors.
+	var formattedErrors []ErrorObject
+	for _, err := range errs {
+		e := marshalError(err)
+		formattedErrors = append(formattedErrors, e)
+	}
+
+	// Write out the serialize errors payload.
+	if err := json.NewEncoder(w).Encode(&ErrorsPayload{Errors: formattedErrors}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// marshalError will serialize the given error as best as possible according to this
+// package's `Error<field>Compatible` interfaces.
+func marshalError(err error) ErrorObject {
+	errorObject := ErrorObject{}
+	if e, ok := err.(ErrorIDCompatible); ok {
+		errorObject.ID = e.GetID()
+	}
+
+	if e, ok := err.(ErrorTitleCompatible); ok {
+		errorObject.Title = e.GetTitle()
+	} else {
+		errorObject.Title = fmt.Sprintf("Encountered error of type: %T", err)
+	}
+
+	if e, ok := err.(ErrorDetailCompatible); ok {
+		errorObject.Detail = e.GetDetail()
+	} else {
+		errorObject.Detail = err.Error()
+	}
+
+	if e, ok := err.(ErrorStatusCompatible); ok {
+		errorObject.Status = e.GetStatus()
+	}
+
+	if e, ok := err.(ErrorCodeCompatible); ok {
+		errorObject.Code = e.GetCode()
+	}
+
+	if e, ok := err.(ErrorMetaCompatible); ok {
+		errorObject.Meta = e.GetMeta()
+	}
+
+	return errorObject
+}
+
+// ErrorsPayload is a serializer struct for representing a valid JSON API errors payload.
+type ErrorsPayload struct {
+	Errors []ErrorObject `json:"errors"`
+}
+
+// ErrorObject is an `Error` implementation as well as an implementation of the JSON API error object.
+// The main idea behind this struct is that you
+// For more information on Golang errors, see: https://golang.org/pkg/errors/
+// For more information on the JSON API spec's error objects, see: http://jsonapi.org/format/#error-objects
+type ErrorObject struct {
+	// ID is a unique identifier for this particular occurrence of a problem.
+	ID string `json:"id,omitempty"`
+
+	// Title is a short, human-readable summary of the problem that SHOULD NOT change from occurrence to occurrence of the problem, except for purposes of localization.
+	Title string `json:"title"`
+
+	// Detail is a human-readable explanation specific to this occurrence of the problem. Like title, this field’s value can be localized.
+	Detail string `json:"detail"`
+
+	// Status is the HTTP status code applicable to this problem, expressed as a string value.
+	Status string `json:"status,omitempty"`
```

Comment:
I'd also recommend making Status a `uint` such that we can directly use the HTTP status code constants defined in https://golang.org/pkg/net/http/#pkg-constants